File: ch/Chess/Piece.cs
using System.Collections.Generic;

namespace Chess
{
    public class Piece
	{
	    public readonly PieceColor Color;
	    public readonly PieceType PieceType;

		public Piece(PieceType pieceType, PieceColor color)
		{
			PieceType = pieceType;
			Color = color;
		}

        public IEnumerable<Location> GetMoves(Location location, Board board) => 
            PieceType.GetMoves(location, board);

	    public override string ToString()
		{
			var c = PieceType == null ? " ." : " " + PieceType;
			return Color == PieceColor.Black ? c.ToLower() : c;
		}
	}

    public static class PieceExtensions
    {
        public static bool Is(this Piece piece, PieceColor color) => 
            piece != null && piece.Color == color;

        public static bool Is(this Piece piece, PieceColor color, PieceType pieceType) => 
            piece.Is(color) && piece.PieceType == pieceType;
    }
}

File: java/src/main/java/ru/kontur/courses/chess/Piece.java
package ru.kontur.courses.chess;

import java.util.stream.Collectors;

public class Piece {
    public final PieceColor color;
    public final PieceType pieceType;

    public Piece(PieceType pieceType, PieceColor color) {
        this.pieceType = pieceType;
        this.color = color;
    }

    public Iterable<Location> getMoves(Location location, Board board) {
        return pieceType.getMoves(location, board).collect(Collectors.toList());
    }

    @Override
    public String toString() {
        var c = pieceType == null ? " ." : " " + pieceType;
        return color == PieceColor.BLACK ? c.toLowerCase() : c;
    }

    public static boolean is(Piece piece, PieceColor color) {
        return piece != null && piece.color == color;
    }

    public static boolean is(Piece piece, PieceColor color, PieceType pieceType) {
        return is(piece, color) && piece.pieceType == pieceType;
    }
}
